src/session.c

```diff
@@ -232,20 +435,93 @@ CK_RV p11prov_get_session(P11PROV_CTX *provctx, CK_SLOT_ID *slotid,
         *next_slotid = CK_UNAVAILABLE_INFORMATION;
     }
 
-    p11prov_ctx_unlock_slots(provctx, &slots);
-
     if (ret != CKR_OK) {
         return ret;
     }
 
-    sess = p11prov_session_new(provctx, id);
-    if (sess == NULL) {
-        return CKR_HOST_MEMORY;
+    ret = CKR_OK;
+
+    /* LOCKED SECTION ------------- */
+    pthread_mutex_lock(&slot->pool->lock);
+    for (i = 0; i < slot->pool->num_sessions; i++) {
+        if (slot->pool->sessions[i]->free) {
+            bool ok;
+            int expected = 1;
+
+            session = slot->pool->sessions[i];
+
+            ok =
+                __atomic_compare_exchange_n(&session->free, &expected, 0, false,
+                                            __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+            if (!ok) {
+                P11PROV_raise(provctx, CKR_GENERAL_ERROR,
+                              "Unexpectd busy session while holding lock");
+                ret = CKR_GENERAL_ERROR;
+            }
+            break;
+        }
     }
-    ret = p11prov_session_open(sess, false, NULL, 0);
+    pthread_mutex_unlock(&slot->pool->lock);
+    /* ------------- LOCKED SECTION */
+
     if (ret != CKR_OK) {
         return ret;
     }
-    *session = sess;
-    return CKR_OK;
+
+    if (session == NULL) {
+        session = p11prov_session_new(provctx, slot);
+        if (session == NULL) {
+            return CKR_GENERAL_ERROR;
+        }
+    }
```

Comment:
Somehow this feels racy. Walking the `session` list is done using the lock held, but it is released before `p11prov_session_new` call. I'd try to make the walking and inserting new object into one batch operation. Especially considering that a potential sleeping would be needed in this place. See the review message for next commit.

Reply:
This is safe p11prov_session_new() locks the pool on its own, and is done only if no free sessions were found.
It is true that a session may come free just as we exit the locked section and before entering p11prov_session_new(), but that changes nothing because if we moved p11prov_session_new() inside the locked block [hard because then we need to detect that and avoid locking inside_new()] then we'd also have created a new session regardless as nothing could free anything in the meanwhile. So bottom line we end up with a new session no matter what we do.

By keeping p11prov_session_new() outside of the locked block we simplify the locked sections and keep them tidy. That was one of guiding principle in changing this code. Reducing locked sections (which means reducing contention) is very important here if people are using Hardware Acceleration or HSMs.